[WIP] Fix py3k

tacaswell opened the pull request.

tacaswell committed:
FIX: python compatibility

 - *.py files updated using moderize
 - hand-modified cpp

tacaswell committed:
FIX: bytes vs string issues

Still some failing tests

tacaswell committed:
FIX: Truncate does not move file pointer

Must move back to beginning via seek.

ref:
  https://docs.python.org/2/library/stdtypes.html#file.truncate
  https://docs.python.org/3.5/library/io.html?highlight=truncate#io.IOBase.truncate

tacaswell committed:
FIX: read file as binary

tacaswell committed:
FIX: e.message has been deprecated from 2.6

tacaswell committed:
FIX: yield bytes, not ints

Iterating over a bytes yields integers, taking length 1 slices results
in 1 byte.

tacaswell committed:
TST: test inputs must be byte literals (not str)

tacaswell committed:
STY: white space changes

tacaswell committed:
MNT: remove pre 2.7 work arounds

tacaswell committed:
BLD: include top-level test module

mdavidsaver commented:
Which python versions are you using/testing against?  Do the unittests still pass?

By and large these look like straightforward mechanical changes to use the six module, so no problem there.  I'll need to get confirmation from Anton that python 2.6 (aka Debian 6) support can be dropped before merging.

arkilic commented:
attn @aderbenev

tacaswell commented:
What is the preferred way to run the units tests?  I was just running `py.test` in the root directory and that seemed to work.  I did all of this under 3.5.

Everything after the first commit was changes made to make that pass.  I suspect that there are still some edge cases of bytes vs unicode floating around in the user facing API.

Is it possible to get travis set up on this project?

aderbenev commented:
We have some machines that still run Debian 6. However, all of them are IOC servers so I don't expect support drop to hurt.

arkilic commented:
:+1:

mdavidsaver commented:
> What is the preferred way to run the units tests? I was just running py.test in the root directory and that seemed to work.

The twisted matrix 'trial' utility.  http://twistedmatrix.com/trac/wiki/TwistedTrial

> I did all of this under 3.5.

For completeness I should also ask for versions of other modules, twisted being the big one.

mdavidsaver commented:
Ok, so it's reasonable to make python 2.7 the minimum supported (aka debian 7).

tacaswell commented:
I think twisted 15.4 (I either did pip install or conda install, I don't remember which off the top of my head, but that is the version both get).

arkilic commented:
I tried the api to pull some data out of LIX archiver appliance using what is installed in the machine (python 2.7) and python 3 compatible version provided by @tacaswell in this pr. Python 3 compatible version fails as the following:
python 2.7: http://nbviewer.ipython.org/gist/arkilic/b41dfe5024f069e3e6c1
python 3: http://nbviewer.ipython.org/gist/arkilic/e63f0ce948b25da6dbb0
The issue ```TypeError: Can't convert 'bytes' object to str implicitly``` is replicated in other boxes as well.

tacaswell commented:
@mdavidsaver I am failing to figure out how to use `trial` to actually run the tests.

mdavidsaver commented:
From the top source directory, or with PYTHONPATH set appropriately, run
"trial <module.name>" which in this case is "trial carchive".
On Nov 20, 2015 11:07 AM, "Thomas A Caswell" <notifications@github.com>
wrote:

> @mdavidsaver <https://github.com/mdavidsaver> I am failing to figure out
> how to use trial to actually run the tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/epicsdeb/carchivetools/pull/7#issuecomment-158461783>.
>

mdavidsaver commented:
Ok, in theory I've enabled https://travis-ci.org/epicsdeb/carchivetools.  Please rebase against the latest and lets see if I checked all the necessary boxes.

You'll also need to add python 3.x to the .travis.yml file and create a new requirements-X.txt file (where X somehow describes your environment).  Obviously the deb7 and deb8 profiles won't build with python 3 and these combinations should be excluded.

mdavidsaver committed:
fixup some py3 issues

mdavidsaver committed:
update travis

mdavidsaver committed:
zope implements is broken in py3

mdavidsaver committed:
unsix _conf

apparently six 1.1.0 circa. debian 7 doesn't handle
the configparser move.

mdavidsaver committed:
mktime/timegm want struct_time or tuple in py3

mdavidsaver committed:
update six in deb7 and protobuf in testing

six 1.1.0 which shipped with debian 7 seems to be not quite complete.
1.8.0 is presently in deb7 backports, so use it for now.

protobuf 2.6.1 doesn't work with py3, try the 3.0 beta.

mdavidsaver commented:
I made an attempt to fix some of these issues, and almost succeeded (https://github.com/mdavidsaver/carchivetools/tree/py3).  The remaining blocker for the CI builder is probably incompatibility between the python stub generated by protoc and the protobuf module.  Working around this will be annoying.  If you have matching versions it will probably work though.

The other issue is what looks like a rounding error in the date handling.  Don't know if this is intrinsic to python3.4 or the travisci environment.  Odd, but I won't lose sleep over 1 microsecond.

https://travis-ci.org/mdavidsaver/carchivetools/builds/96897589

tacaswell commented:
There is a change in the rounding between python 2 and python 3 (https://docs.python.org/3/whatsnew/3.0.html#builtins)  to be less biased.  If you want it to be the same in both python 2 and 3, `np.round` implements the new strategy on both.

tacaswell commented:
Pulled your commits on top of this branch just for house keeping.

mdavidsaver commented:
cf. #10
